### PR TITLE
Add hook to optionally run prestart commands in GoldenGate 21c

### PR DESCRIPTION
**Pull request by tmelhiser**
This allows the container to optionally execute prestart scripts or commands in the container to get it ready before starting the goldengate processes.

### CODE REVIEW

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:141]: I think this function could consist of a single one-line statement:

```sh
eval "$(RUN_COMMAND}" || abort "${RUN_COMMAND} did not complete successfully!"
```

An empty value for `RUN_COMMAND` becomes a no-op.

If `RUN_COMMAND` is terminated with `&`, the task is executed as a background job. That would also allow for more complex background tasks run with `nohup` and similar wrappers.

**Review by sbalousek — CHANGES_REQUESTED** [review]
Hi @tmelhiser,

I have one comment regarding commands executed in the background (see later), but more importantly, would it make more sense to follow the pattern established for OracleDatabase? That project looks for two directories, `${ORACLE_BASE}/scripts/setup` and `${ORACLE_BASE}/scripts/startup`. The scripts are run before and after the DBMS starts.

If we followed the same pattern, the container for OGG would use a single volume mount for any number of configuration and setup scripts.

- Steve

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (tmelhiser) [reply OracleGoldenGate/21c/bin/deployment-main.sh:141]: The mounted volume pattern is nice. I think this is the route I will approach as it gives a known set of tested scripts that are run.

I was not a great fan of the arbitrary RUN_COMMAND approach to begin with, due to the ability to run any arbitrary code in the container as a privileged user.  I avoided the eval statement even more so, due to it's expansion allowing for even more complex execution patterns.

The directory/mounted volume approach appears to force a foreground linear lexicographical approach.

Is there any desire to allow for asynchronous background processes to be started as well?

This could be achieved with an intuitive directory structure:

| - hooks
&nbsp;&nbsp;&nbsp;&nbsp;| -- background
&nbsp;&nbsp;&nbsp;&nbsp;| -- prestart

Although possible, I did not propose an exit/cleanup folder.  This enhancement could be augmented in the future, but the current signal handling and health monitoring would need a good bit of rework to ensure OGG was in the correct state while not having the container terminated/restarted by the container management system prior to the scripts finishing.

Additional feedback would be appreciate before proceeding.

Thanks again.

Comment (tmelhiser) [reply OracleGoldenGate/21c/bin/deployment-main.sh]: Updated to use the directory pattern ${OGG_HOME}/scripts/setup and ${OGG_HOME}/scripts/startup

The backgrounding issue can be handled on a per script basis inside the script itself, rather than as an arbitrary flag in the config.  Allowing user determination on what runs in the background and in what order.

The scripts will be run lexicographically and will recursively span subdirectories.
The order preference of files is determined only by the lexicographical order on a per directory basis.
ie: 
> 02-sanity.sh
> 01-connectivity-validation\
>>> 02-test-hostA.sh 
>>>01-test-hostZ.sh 

will be run in the lexicographic order:
1. 01-test-hostZ.sh
2. 02-test-hostA.sh
3. 02-sanity.sh

SQLPlus is dynamically located and .sql files can be run.
Scripts in */setup will be run prior to any other startup sequence
Scripts in */startup will be run after ogg/nginx successfully start

File: OracleGoldenGate/21c/Dockerfile
Comment (sbalousek) [inline OracleGoldenGate/21c/Dockerfile:17]: This should be handled by the file mode stored in Git. Did you run into problems on Windows maybe?

File: OracleGoldenGate/21c/README.md
Comment (sbalousek) [inline OracleGoldenGate/21c/README.md:69]: The convention established in the `OracleDatabase` project is a pre-coded location of either:
- `"$ORACLE_BASE"/scripts/setup`
- `"$ORACLE_BASE"/scripts/startup`

I would expect something similar here, with:
- `"$OGG_HOME"/scripts/setup`
- `"$OGG_HOME"/scripts/startup`

Is there really any value in allowing control over the script directory location when a volume mount already controls the content of those directories?

File: OracleGoldenGate/21c/README.md
Comment (sbalousek) [inline OracleGoldenGate/21c/README.md:71]: It might be better to handle errors from within the scripts. Perhaps if there was a generic `abortContainer.sh` command that terminated the container with an error message called from the user script?

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:32]: These `mkdir` commands can go in the `Dockerfile`

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:36]: Since OGG is not the DBMS, any SQL processing should be handled entirely in the user scripts.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:174]: We can't execute SQL*Plus from here because the database is likely not in the same container. When running in another container, credentials are needed, etc.

Because of the inherent complexities, I would like to see support for `*.sql' files removed.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:168]: I would vote for executing anything with the executable bit set. That way, Python scripts (or whatever) could also be run. But, the Database only supports `*.sh` files, so maybe we should stick to just those.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:186]: Sleep?  I no sleep!
- Gollum

What are you thinking here with the pause?

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:226]: Er. Why kill the Service Manager here?

File: OracleGoldenGate/21c/install-deployment.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/install-deployment.sh:85]: These should be in `Dockerfile`, and any permissions mismatches between the mounted scripts volume will need to be addressed in the host environment.

**Review by sbalousek — CHANGES_REQUESTED** [review]
Hi @tmelhiser,

I like the direction this is heading in. I have quite a few questions and one major correction with regard to the script directory location.

- Steve

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (tmelhiser) [reply OracleGoldenGate/21c/bin/deployment-main.sh:170]: I like the idea of sourcing the file, since many times values derived from one script can be reused/passed to subsequent ones.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (tmelhiser) [reply OracleGoldenGate/21c/bin/deployment-main.sh:168]: I really like the idea of checking the execute bit.  The issue I worry about there is the oddities we see in git check ins where those aren't stored correctly.

I will defer to your call here.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (tmelhiser) [reply OracleGoldenGate/21c/bin/deployment-main.sh:36]: I will drop off the sqlplus hook.  I assume this also means dropping the script execution of *.sql files?

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (tmelhiser) [reply OracleGoldenGate/21c/bin/deployment-main.sh:32]: This kind of goes back to the question of "Do we allow user defined locations of where the *_USER_SCRIPTS are located".  If we allow user definition, this is a nice safeguard preventing errors in the belly of the script due to missing dirs.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [reply OracleGoldenGate/21c/bin/deployment-main.sh:170]: Yes, but:

- The `source` command can be used from those scripts.
- Using `source` here potentially corrupts the execution environment

Since the `source` option is used by the DBMS team, we should use it here, too.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [reply OracleGoldenGate/21c/bin/deployment-main.sh:168]: Let's stick to the DBMS model and only process `*.sh` files via `source` (or `.`)

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [reply OracleGoldenGate/21c/bin/deployment-main.sh:32]: Since the DBMS containerized image does not allow a user-defined location for scripts, I would like to use a fixed location here, too.

File: OracleGoldenGate/21c/README.md
Comment (tmelhiser) [reply OracleGoldenGate/21c/README.md:71]: If I understand correctly, the abortContainer.sh would be expected to be called by the third party scripts as part of their error handling?  ie: their trap conditions would have to implement handing control to abortContainer.sh, which would then have to handle determining whether to abort or hand control back to the originating deployment-main.sh or continue with the existing thrid party script?

This seems like a pretty complex scripting model for a large majority of the user base to comprehend.

If simplification and avoidance confusion by the ABORT_ON_USER_SCRIPT_ERROR is the goal,  I would suggest the same as you -- any script non-zero exit aborts the boot; forcing the users to handle errors in their scripts.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (tmelhiser) [reply OracleGoldenGate/21c/bin/deployment-main.sh:226]: The OGG primary instance never receives the os level termination signal.  The original code trying to trap the OGG PID is actually only trapping the PID of the tail command.

Since the PID of the OGG instance is created inside the python script, it must be gathered there.  In order to issue the shutdown of OGG when the main PID 1 is signaled to exit by the orchestration layer (docker/k8s/etc) the trap then kills the child processes.  The PID of the OGG instance was never really being notified.

This correctly sends the exit signal to the OGG process and it shutsdown cleanly, rather than waiting the container timeout period and forcefully killing all remaining PIDs.

File: OracleGoldenGate/21c/install-deployment.sh
Comment (tmelhiser) [reply OracleGoldenGate/21c/install-deployment.sh:85]: I don't think I changed anything in this area other than formatting?  The extra space between the -p and the ${OGG_HOME}.

Probably just me being pedantic about the spacing... if you prefer the space, I can put it back.

File: OracleGoldenGate/21c/Dockerfile
Comment (tmelhiser) [reply OracleGoldenGate/21c/Dockerfile:17]: No, this was built on a linux machine.

Comment (sbalousek) [reply OracleGoldenGate/21c/Dockerfile:17]: OK. Well, the `install-prerequisites.sh` script is required. I believe the changes to `Dockerfile` should be reverted,

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [reply OracleGoldenGate/21c/bin/deployment-main.sh:186]: In Docker, the logs for a container are available until the container is destroyed.

Regardless, if it's specific to your environment, it should be in a user script. That would follow the behavior in thee DBMS container.

File: OracleGoldenGate/21c/install-deployment.sh
Comment (sbalousek) [reply OracleGoldenGate/21c/install-deployment.sh:85]: Unless reformatting with a purpose, whitespace-only changes should be avoided.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [reply OracleGoldenGate/21c/bin/deployment-main.sh:226]: I'm not sure what you mean by 'OGG primary instance.'  As far as I know, all processes in a Docker container are sent the SIGTERM signal when the container is terminated. Service Manager, in particular, shuts down pretty quickly when it receives SIGTERM.

I would like to better understand which processes are not responding in a timely fashion to SIGTERM.

Anyway, this change is not really part of enable user scripts, so it should be in a separate PR at a minumum.

File: OracleGoldenGate/21c/README.md
Comment (sbalousek) [reply OracleGoldenGate/21c/README.md:71]: The `abortContainer.sh` script was just a thought and not something I expected to be part of this PR. IN the DBMS scripts (see `OracleDatabase/SingleInstance/dockerfiles/21.3.0/runUserScripts.sh`), there is no error handling.

Although I am not fond of that naive approach, I believe that whatever is implemented for OGG should be as similar as possible to what is already being done for the DBMS images. The reason for that belief is consistent implementation and behavior.

**Review by abhijitvk — COMMENTED** [review]
@sbalousek : All the requested changes are completed and are ready for your re-review. Would appreciate your time to get this reviewed and provide feedback and /or merge at the earliest convenience.

File: OracleGoldenGate/21c/install-deployment.sh
Comment (abhijitvk) [reply OracleGoldenGate/21c/install-deployment.sh]: reverted the whitespace changes

**Comment by abhijitvk** [comment]
@sbalousek Requesting feedback on the PR and merge if all the earlier comments have been addressed. Thank you!

File: OracleGoldenGate/21c/Dockerfile
Comment (sbalousek) [reply OracleGoldenGate/21c/Dockerfile]: The use of `chmod` is not required. All shell scripts and other executables already have the executable bit set in this Git repository. These changes can be reverted.

File: OracleGoldenGate/21c/bin/deployment-init.py
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-init.py:136]: Although a nice addition, these changes are unrelated to the focus of this PR, which is startup scripts. Please open a separate issue regarding Basic/Digest authentication and create a PR for that issue separately,

File: OracleGoldenGate/21c/README.md
Comment (sbalousek) [reply OracleGoldenGate/21c/README.md:69]: This issue needs resolution.

**Review by sbalousek — CHANGES_REQUESTED** [review]

**Review by sbalousek — CHANGES_REQUESTED** [review]
Style changes are not desirable. Please remove.

**Review by sbalousek — CHANGES_REQUESTED** [review]
These changes have nothing to do with the issue of running scripts at Deployment startup.

File: OracleGoldenGate/21c/bin/deployment-init.py
Comment (abhijitvk) [reply OracleGoldenGate/21c/bin/deployment-init.py:136]: While trying to run the container each time I was seeing the error stacktrace as below on the Linux machine. Basic auth is supported by the API calls and should work absolutely fine. Fallback would be the digest authentication. I will create a separate PR for this one if that makes more sense.

`2022-06-01T13:20:23.070Z | INFO    | ALL DONE (0 errors).
Traceback (most recent call last):
  File "/usr/local/bin/deployment-init.py", line 265, in <module>
    sys.exit(main())
  File "/usr/local/bin/deployment-init.py", line 260, in main
    create_ogg_deployment()
  File "/usr/local/bin/deployment-init.py", line 251, in create_ogg_deployment
    reset_service_configuration(serviceName)
  File "/usr/local/bin/deployment-init.py", line 166, in reset_service_configuration
    config = get_service_config(serviceName)
  File "/usr/local/bin/deployment-init.py", line 140, in get_service_config
    response = get_requests_session().get(url, headers=rest_call_headers, auth=get_digest_authentication())
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 555, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 662, in send
    r = dispatch_hook('response', hooks, r, **kwargs)
  File "/usr/lib/python3.9/site-packages/requests/hooks.py", line 31, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)
  File "/usr/lib/python3.9/site-packages/requests/auth.py", line 267, in handle_401
    prep.headers['Authorization'] = self.build_digest_header(
  File "/usr/lib/python3.9/site-packages/requests/auth.py", line 185, in build_digest_header
    HA1 = hash_utf8(A1)
  File "/usr/lib/python3.9/site-packages/requests/auth.py", line 148, in md5_utf8
    return hashlib.md5(x).hexdigest()
ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS `

File: OracleGoldenGate/21c/Dockerfile
Comment (abhijitvk) [reply OracleGoldenGate/21c/Dockerfile]: I took the chmod out and if needed can be done via Dockerfile add-on or user script

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (abhijitvk) [reply OracleGoldenGate/21c/bin/deployment-main.sh:166]: replaced with the source for processing *.sh files

File: OracleGoldenGate/21c/README.md
Comment (abhijitvk) [reply OracleGoldenGate/21c/README.md]: The issue has been resolved now. Not allowing any more control over the scripts directory as the volume mount controls it.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (abhijitvk) [reply OracleGoldenGate/21c/bin/deployment-main.sh]: Updated to be the fixed location

**Comment by abhijitvk** [comment]
@sbalousek Feedback changes are all completed and ready for review + merge. Appreciate your quick response

File: OracleGoldenGate/21c/README.md
Comment (abhijitvk) [inline OracleGoldenGate/21c/README.md:146]: @sbalousek  Adding this new section in README for more clarity

**Comment by abhijitvk** [comment]
Are these changes ready to be merged? @sbalousek

File: OracleGoldenGate/21c/README.md
Comment (sbalousek) [inline OracleGoldenGate/21c/README.md:147]: Please back-tick quote literals like `.sh` and directory names

File: OracleGoldenGate/21c/README.md
Comment (sbalousek) [inline OracleGoldenGate/21c/README.md:149]: Backtick quote, please

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:145]: `${SCRIPTS_ROOT}` will never be empty

File: OracleGoldenGate/21c/install-deployment.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/install-deployment.sh:89]: or:
```.sh
mkdir -p '${OGG_HOME}/scipts/"{setup,startup}
chown -R ogg:ogg "${OGG_HOME}/scripts"
```

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:208]: Is this really needed?

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (sbalousek) [inline OracleGoldenGate/21c/bin/deployment-main.sh:30]: Can these two variables be reduced to simply `OGG_DEPLOYMENT_SCRIPTS`?

**Review by sbalousek — CHANGES_REQUESTED** [review]
I feel like these changes are very close. In addition, I believe the `Dockerfile` should include the environment variables for the scripts directory and export that directory as a volume.

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (abhijitvk) [reply OracleGoldenGate/21c/bin/deployment-main.sh]: Combined into a single variable `OGG_DEPLOYMENT_SCRIPTS`

File: OracleGoldenGate/21c/install-deployment.sh
Comment (abhijitvk) [reply OracleGoldenGate/21c/install-deployment.sh]: replaced for loop with suggested mkdir single execution

File: OracleGoldenGate/21c/README.md
Comment (abhijitvk) [reply OracleGoldenGate/21c/README.md]: added Backtick quote

File: OracleGoldenGate/21c/README.md
Comment (abhijitvk) [reply OracleGoldenGate/21c/README.md]: added Backtick quote

File: OracleGoldenGate/21c/bin/deployment-main.sh
Comment (abhijitvk) [reply OracleGoldenGate/21c/bin/deployment-main.sh:142]: It won't be assuming that the function `run_user_scripts` is invoked with a non-empty value. Additional check should be OK i guess.

**Comment by abhijitvk** [comment]
All the suggested changes are completed. Should be ready for merge at this time. let me know for any other feedback

**Comment by abhijitvk** [comment]
> Thank you for all your efforts on this. I hope all my change requests did not drive you crazy!

Sure thing! Thanks for your valuable feedback. Is there a process before merging into the main branch?

**Pull request merged**
